rfc/rfc-51/rfc-51.md

```diff
@@ -64,69 +65,72 @@ We follow the debezium output format: four columns as shown below
 
-| key | default  | description |
-| --- |---| --- |
-| hoodie.datasource.query.type | snapshot | set to cdc, enable the cdc quey mode |
-| hoodie.datasource.read.start.timestamp | -        | requried. |
-| hoodie.datasource.read.end.timestamp | -        | optional. |
+For examples:
+- `INSERT` operation will maybe create a list of new data files. These files will be treated as ADD_BASE_FILE;
+- `DELETE_PARTITION` operation will replace a list of file slice. For each of these, we get the cdc data in the `REPLACED_FILE_GROUP` way.
 
+## When `supplemental.logging=false`
 
-### CDC File Types
+In this mode, we minimized the additional storage for CDC information. 
 
-Here we define 5 cdc file types in CDC scenario.
+- When write, the logging process is similar to the one described in section "When `supplemental.logging=true`", just that only change type `op` and record key are persisted.
+- When read, changed info will be inferred on-the-fly, which costs more computation power.
 
-- CDC_LOG_File: a file consists of CDC Blocks with the changing data related to one commit.
-  - when `hoodie.table.cdc.supplemental.logging` is true, it keeps all the fields about the change data, including `op`, `ts_ms`, `before` and `after`. When query hudi table in cdc query mode, load this file and return directly.
-  - when `hoodie.table.cdc.supplemental.logging` is false, it just keeps the `op` and the key of the changing record. When query hudi table in cdc query mode, we need to load the previous version and the current one of the touched file slice to extract the other info like `before` and `after` on the fly.
-- ADD_BASE_File: a normal base file for a specified instant and a specified file group. All the data in this file are new-incoming. For example, we first write data to a new file group. So we can load this file, treat each record in this as the value of `after`, and the value of `op` of each record is `i`.
-- REMOVE_BASE_FILE: a normal base file for a specified instant and a specified file group, but this file is empty. A file like this will be generated when we delete all the data in a file group. So we need to find the previous version of the file group, load it, treat each record in this as the value of `before`, and the value of `op` of each record is `d`.
-- MOR_LOG_FILE: a normal log file. For this type, we need to load the previous version of file slice, and merge each record in the log file with this data loaded separately to determine how the record has changed, and get the value of `before` and `after`.
-- REPLACED_FILE_GROUP: a file group that be replaced totally, like `DELETE_PARTITION` and `INSERT_OVERWRITE` operations. We load this file group, treat all the records as the value of `before`, and the value of `op` of each record is `d`.
+Detailed inference algorithms are illustrated in this [design document](https://docs.google.com/document/d/1vb6EwTqGE0XBpZxWH6grUP2erCjQYb1dS4K24Dk3vL8/).
```

Comment:
this google doc and the algorithm to extract cdc when `supplemental.logging=false` is not same.
both the two algorithms will load the previous version and the current version of a file group to two maps which the key is the record key and the value is the whole record. But in case that we query cdc when `supplemental.logging=false`, just iterator the cdc_log_file and use the record key to get the `before` and `after` values in ad-hoc way. The algorithm in this doc is like a hash-map join to extract cdc.
Assume that the previous version has M records, the current one has N records, and there are R records updated/deleted/inserted. The algorithm when `supplemental.logging=false` spends O(R), the one in doc spends O(M+N). If i'm wrong, correct me please.

Reply:
Overall I think there is a difference in the design proposed in HUDI-3478 and in HUDI-1771

HUDI-1771 is a simple elegant design but as mentioned in [comment](https://github.com/apache/hudi/pull/5436#issuecomment-1206165913) - I am more inclined to the RFC-51 design that tracks the op and row_keys in the CDC log block and the reader doing a hash based join with this and the required file slices to compute the CDC stream.

This will be a batched row_keys lookup and we should be efficient in doing that in all engines and the complexity is close to O(updates) and that is very reasonable rather than O(data accessed) like @YannByron mentions here.

Reply:
yes the original algorithm in the doc assumes no cdc data (key and op) is logged hence it won't be as optimized as when logging key and op. it should be updated accordingly, at least on this PR to align with the case of logging key and op.

Reply:
maybe `read_cdc_log_file.png` is enough to explain this logic. 
we can update this png if we have already used `hoodie.table.cdc.supplemental.logging.mode` instead of `hoodie.table.cdc.supplemental.logging`.

Reply:
Have updated the PR for this accordingly.